cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/S3BatchSource.java

```diff
@@ -0,0 +1,263 @@
+  private static final Logger LOG = LoggerFactory.getLogger(S3BatchSource.class);
+  private S3BatchConfig config;
+
+  private static final Schema DEFAULT_SCHEMA = Schema.recordOf(
+    "event",
+    Schema.Field.of("ts", Schema.of(Schema.Type.LONG)),
+    Schema.Field.of("body", Schema.of(Schema.Type.STRING))
+  );
+
+  public S3BatchSource (S3BatchConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    //will create the table if it doesn't exist already
+    if (config.timeTable != null) {
+      pipelineConfigurer.createDataset(config.timeTable, KeyValueTable.class, DatasetProperties.EMPTY);
+    }
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    //calculate date one minute ago, rounded down to the nearest minute
+    prevMinute = new Date(System.currentTimeMillis() - TimeUnit.MINUTES.toMillis(1));
+    Calendar cal = Calendar.getInstance();
+    cal.setTime(prevMinute);
+    cal.set(Calendar.SECOND, 0);
+    cal.set(Calendar.MILLISECOND, 0);
+    prevMinute = cal.getTime();
+    currentTime = new SimpleDateFormat("yyyy-MM-dd-HH-mm").format(prevMinute);
+
+    Job job = context.getHadoopJob();
+    job.getConfiguration().set("fs.s3n.awsAccessKeyId", config.accessID);
+    job.getConfiguration().set("fs.s3n.awsSecretAccessKey", config.accessKey);
+    if (config.fileRegex != null) {
+      job.getConfiguration().set("input.path.regex", config.fileRegex);
+    }
+    job.getConfiguration().set("input.path.name", config.path);
+    if (config.timeTable != null) {
+      table = context.getDataset(config.timeTable);
+    }
+
+    FileInputFormat.setInputPathFilter(job, S3Filter.class);
+    FileInputFormat.addInputPath(job, new Path(config.path));
+  }
+
+  @Override
+  public void transform(KeyValue<LongWritable, Text> input, Emitter<StructuredRecord> emitter) throws Exception {
+    StructuredRecord output = StructuredRecord.builder(DEFAULT_SCHEMA)
+      .set("ts", System.currentTimeMillis())
+      .set("body", input.getValue().toString())
+      .build();
+    emitter.emit(output);
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSourceContext context) {
+    if (succeeded && table != null && "timefilter".equals(config.fileRegex)) {
+      table.write("lastTimeRead", currentTime);
+    }
+  }
+
+  /**
+   * Filter class to filter out filenames in the input path.
+   */
+  public static class S3Filter extends Configured implements PathFilter {
+    private boolean useTimeFilter;
+    private Pattern regex;
+    private String pathName;
+
+    @Override
+    public boolean accept(Path path) {
+      String filename = path.toString();
+      //InputPathFilter will first check the directory if a directory is given
+      if (filename.equals(pathName) || filename.equals(pathName + "/")) {
+        return true;
+      }
+
+      if (useTimeFilter) {
```

Comment:
I think if you invert this condition, you can then put lines 197-198 in the if block.
And then lines 163-193 can be unindented. Confirm that it would be equivalent.
In my opinion, at least, the less indentation/nesting, the better.